### ISSUE TITLE
augmented modifier for members?

### ISSUE THREAD

**Issue by johnniwinther** [dart-lang/language#3580]
When generating macros in three phases, the situation can occur where the second phase adds unimplemented members that are implementated in phase 3. For instance we might have
```Dart
@AddMemberMacro()
class Class {}
```
from which the second phase creates
```Dart
augment class Class {
  @AddMemberBodyMacro()
  int method();
}
```
with the intent for the third phase to create
```Dart
augment class Class {
  augment int method() {
    print('Hello');
    return 42;
  }
}
```
to complete the class.

A problem arises from this because outlines (aka summaries) are created after the second phase, and with that, the checking of the class hierarchy. For the example above this leads to an error being reported that the non-abstract `Class` has an abstract member `method`.

Currently macros have to avoid this either by generating a dummy body:
```Dart
augment class Class {
  @AddMemberBodyMacro()
  int method() => throw UnimplementedError();
}
```
or mark the member as `external`:
```Dart
augment class Class {
  @AddMemberBodyMacro()
  external int method();
}
```

Both have the problem that if the macro fails to add the intended body, no error is reported. (The support for external members is very much dependent on the backend so we might get an error at compile-time or runtime.)

I suggest that we add an `augmented` modifier to use in this scenario:
```Dart
augment class Class {
  @AddMemberBodyMacro()
  augmented int method();
}
```
In the class hierarchy checking this should be handled as if the method being non-abstract. After phase 3, a check should be made that the member was in fact augmented to have a body.

A suggested alternative is to simply assume the "augmented" modifier on abstract members added in phase 2, but this poses a problem for when the member *was* intended to be abstract (adding members to interface classes for instance) or when the member was intended to be non-abstract but was added to an abstract class. We wouldn't be able to tell these apart without the explicit modifier.

**Comment by johnniwinther** [comment]
cc @jakemac53

**Comment by jakemac53** [comment]
> It's possible, but I also worry that it will make augmentation more verbose.

Yes, this is a definite worry and why this keyword doesn't exist today.

Augmentations can add entirely new declarations, without anything special on the original class. It feels weird if you have to put something special on members, when augmenting those is probably less unexpected if anything (from a user perspective).

It sounds like maybe we have to run phase 3 during summary generation?

**Comment by scheglov** [comment]
In the analyzer we run all 3 phases during linking summaries, because even the definitions of the macro-generated library augmentation are based on elements, and so the whole code of the augmentation, including bodies, is a part of the linked summary.

So, for the analyzer we don't need this `augmented` promise, if it is augmented, we will know it.

**Comment by lrhn** [comment]
If this is _only_ a macro issue, it should probably not be a language feature, just a macro code generation API feature.
When adding the step 2 augmentation member, it could be required to specify whether there will be a step 3 body or not.
I'm not familiar with the API used to add members, but would something like that be possible?

Would it be useful for step 1 as well, having to say whether there will be more in later steps, or if this is it?

**Comment by jakemac53** [comment]
@johnniwinther what makes the macro case different from a user writing this?

```dart
class Thing {
  @ProvideDefinitionInPhase3()
  void doSomething();
}
```

We want that to be possible as well, as long as the macro fills in a body.

**Comment by jakemac53** [comment]
> Can we tell by the interface of the macro that it is going to generate a body?

Not definitively, no. There are two issues:

- The macro could exit early, it isn't obligated to actually emit any code.
- You could also have macros on a surrounding scope (class, library), which are filling in this implementation. But there is no way to know which things they will be filling in.

So, you could tell whether "potentially" a macro might fill it in, but not for sure whether it will.

**Comment by lrhn** [comment]
The more I think about this, and talk to @johnniwinther, the more I like the idea of declaring a non-augmenting declaration as "non-abstract", even if it looks abstract.

It's not just for macros. If I split my class into several parts using augmentations, then the top-level declaration of
```dart
abstract class Foo {
  State initializeState();
}
```
may look like `initialize` is abstract, but I know that it is filled in by an augmentation in a sub-part somewhere.

I can write 
```dart
  State initializeState() => throw UnimplementedError("Replaced by augmentation");
```
but if I make a mistake and don't actually augment it, that turns my programming error into a runtime error.

So, how about this completely outrageous strawman:
```dart
 State initializeState() TODO;
```
We make `TODO` a contextual keyword, which we can use where a method body could otherwise be (where an `async` or `sync*` contextual keyword could occur, just not followed by a real body.

This is a method without a body, but it's a non-abstract method without a body, so it's a compile-time error if there is no augmentation in the program that adds a body to the method.

You don't have to write the `TODO`. An augmentation can still augment an abstract declaration with a body (unless that breaks summaries or something).

Less strawman-y ideas:
```dart
  State initializeState() =>;  // Has "no body". The sequence `=>;` is recognized specially, just like `TODO;`.
```
or
```dart
  State initializeState() augment; // Meaning "augment here!"
```
or
```dart
  State initializeState() late; // Meaning "body supplied later!"
```
The last one is probably a bad(der) idea, since it mixes the concepts of runtime lateness with compile-time lateness.

It's harder to find syntax for variable initializers. Take:
```dart
  int? foo;
```
which is augmented with an initializer later. 
```dart
 int? foo TODO;
```
doesn't work, because it's ambiguous for `final foo TODO;` (the type being optional means we have to be able to determine the parsing by the token after the name. We can't use a context-sensitive keyword or built-in identifier there.)
```dart
 int foo =;
```
looks more promising, like `=>;` it uses the starting syntax of the thing that's omitted, then omits the content.

_(If we do nothing, users can always choose to do `Never get TODO => throw UnimplementedError("Not augmented");` and `State initializeState() => TODO;`, with all the problems that entail, but with some actual documentation effect.)_

**Comment by jakemac53** [comment]
Could we allow something like a `!abstract` keyword, to opt out of the implicit abstractness?

**Comment by lrhn** [comment]
It's very, very handy to just write `int foo();` and not have to do the `abstract int foo();` redundancy.
Except that with augmentations it's no longer redundant. :disappointed: 

I'd really prefer to not force everybody to write `abstract` in front, and migrate the world, so having some way to say "this member isn't really abstract, it just looks like it, but an augmentation will come along and fix it" seems useful.

I don't think you should *have* to use that either, it's mostly for documentation and sanity checking.
Non-macro augmentations can always see all the augmentations, so they know whether there is a body or not.

It'd be a marker that makes it a compile-time error to *not* have a body at the end of augmentation. Like (a langauge feature like) `@override` makes it an error to not be an override, but if you don't write anything, it's still fine.
And then macros can check for the marker, and add them in phase 1 or 2 if they intend to add a body. And we might be able to make it an error for a macro to add a body if it hasn't declared that it wanted to. That's macro API stuff.

The final proof of validity is still that the complete program compiles. If it ends up being valid, even without the needed information, that's nice. (If modular compilation works, obviously.)

**Comment by TekExplorer** [comment]
i actually really like how the unimplemented implementation idea looks. by defining a body, but going _sike_ and not actually giving one, we indicate that there _should_ be something here. after all, its a compile-time error to have nothing there right? so if we allow augmentations to ignore that while processing macros, the analyzer can say, "hold up, this is a concrete thing - why is there no substance?" and warn you.

syntactically, "aborted bodies" (a nice and friendly formal name for this idea) tell the programmer that there's something particular going on there.

we could just make that the syntax for users in general.
in a class, `int? val;` and `int? val =;` can have important meaning.
for instance:
```dart
class Foo {
  @myMacro
  int? bar1; // implies initialization to null. might be an error
  @myMacro
  int? bar2 =; // explicitly states that the macro will handle initialization, and we _dont_ initialize to null (requests augmentation)
}
```
and it would only really be necessary when its ambiguous.
the following is functionally very similar since we could mark things as abstract, but didn't, meaning there should be something here
```dart
class Foo {
  @myMacro
  int? bar1(); // error, not marked abstract. seeing the macro, we can assume it should be filled in. (specify a body, mark it abstract, or augment)
  @myMacro
  int? bar2() =>; // error. body missing. (please augment [specifically]) explicitly showing that the body should be filled in

  int? get bar3 =>; // works even without the macro, because it only cares about augmentation.
}
```
we could then consider this case to be such that `function();` == `function() =>;` as long as they arent abstract

**Comment by munificent** [comment]
The constraints on this issue are different now that macros are out of the picture. Without macros and an introspection API, there's no way to see a library before its augmentations have been applied. So there's no longer a potential ambiguity between "abstract function" and "non-abstract function that will get a body by augmentation".

But I do think there are still two issues:

### Omitting bodies where they currently can't be omitted

When looking at how json_serializable would use augmentations, I ran into a case where it would be good if a user could declare a top-level variable or getter and have the augmentation provide the initializer or body.

A top-level variable without an initializer works in this case, but I'm sure there are others where a user wants to declare a top-level function or getter and have an augmentation provide a body.

Currently, the grammar doesn't allow that:

```
topLevelDeclaration ::= classDeclaration
  ...
  | functionSignature functionBody
  | getterSignature functionBody
  | setterSignature functionBody
```

Every top-level function must have a body. We could of course just allow empty bodies at the top level like we do inside classes for abstract members, but I'm not sure if that's what we want. Which leads to...

### Confusion about which members are abstract versus filled in by an augmentation

Let's say you're looking at a library you don't maintain. You see:

```dart
abstract class A {
  void foo();
  void bar();
}
```

You want to subclass A. Do you have to override `foo()`, `bar()`, both, or neither? If you dig around in the part files (which are really implementation details you shouldn't have to get into) you may eventually discover:

```dart
augment abstract class A {
  augment void bar() {
    ...
  }
}
```

OK, so it turns out `bar()` wasn't abstract after all.

This doesn't seem like great usability to me. I think there should be an explicit marker of some kind on the declaration to mean "not abstract, but the implementation's just not right here". 

## What are the rules around this marker?

Assuming we have some kind of marker for "initializer or body will be filled in by an augmentation", what rules does the compiler have about when you can/must use it? Can you augment an actually-abstract member with a body (and thus making it not abstract)? Is it an error to augment a member that isn't marked as "going to be augmented"?

Here's a pitch (which I think mostly agrees with [Lasse's earlier comment](https://github.com/dart-lang/language/issues/3580#issuecomment-1905992286):

### Incomplete and/or abstract functions

If the function (or method, getter, setter, operator, etc.):

*   Has no body, and:

    *   Is inside an `abstract` class or mixin. This is fine. We treat "no body or other marker" as unambiguously meaning "explicitly declares an abstract member".

        It is an **error if this member is augmented by something that adds a body.** The member is declared to be abstract and augmentations don't generally change the properties of the original declaration. (For example, a factory constructor can't be augmented to become generative, etc.)

        You can augment it with another abstract method, maybe to add metadata annotations, but that's it.

    *   Is in a context where you can't declare an abstract member. **Error.** Since "no body or other marker" means "declare abstract", it's an error to try to do that where you can't have an abstract member.

        Adding an augmentation doesn't help. The code is already erroneous by declaration: it has a declared property that isn't valid in its context.

        This means that we don't need to change the language grammar to allow "abstract top-level functions". There's no such thing. There are just incomplete ones, which must be marked as such.

*   Has a body, and:

    *   Isn't augmented. This is fine, of course. It's already complete!

    *   Is augmented. This is also OK. Augmentations are allowed to replace or wrap functions without the original function having to explicitly permit it. To someone reading the original declaration, they see a function that can be called and what they see is true. The augmentation may change how the *implementation* of that function behaves, but the user-facing API they see is unchanged. It's a callable thing before augmentation and a callable thing after.

*   Has an "incomplete" marker, and:

    *   Isn't augmented. This is an error. **The marker means "this must be augmented" so it's an error for it not to be.**

    *   Is augmented. Fine. The incomplete marker is satisfied. Note that the augmentation must provide a body (or itself be marked incomplete and further augmented by something that eventually fills in a body). It can't be body-less and abstract since that would change the "is abstract" property of the original declaration (from "no" to "yes").

        It is an error for the augmenting function to use `augmented` since there is no original function body to call back into (unless there is some other augmentation in the middle that fills in a body).

In short:

*   The original declaration determines the abstract-ness of the function, and no augmentation can change that.

*   Some kind of "incomplete" marker exists to let you define a concrete function without providing a body.

*   It's an error for an incomplete member to not eventually be augmented with a body.

*   A member can't be both abstract and incomplete. Incomplete means "this will have a body", which is the opposite of "abstract".

### Incomplete variables

We don't have the same conceptual ambiguity around "abstract" and "incomplete" variables that we do with functions. There *are* abstract instance fields, but they have an explicit `abstract` modifier. So a variable declaration is always declaring a concrete variable.

I don't think there's much of a usability problem either. From an API consumer perspective, the *value* of a variable is mostly an implementation detail. So if they see a variable declaration like:

```
final int x;
```

They might not know its value and they might be momentarily confused that a `final` non-nullable variable has no initializer, but they still know what they can do with it. So I don't see much need to support an explicit incomplete modifier for variables.

### Incomplete marker syntax

So how should a function be marked? We already know that the function is concrete, so the marker doesn't change the *properties* of the function. It's really more a description of the function's *implementation* (or momentary lack thereof). In other words, it's about the function's *body*.

So I like the idea of it being a new kind of function body form. Using `...` is very cute but I suspect it's a little *too* cute and confusing with spreads. Using `TODO` is *very* funny, but sends the wrong signal. The function is done. It's just not done *here*.

Using `augmented` is obvious, but I worry that it's confusing. We also use `augmented` inside augmenting functions to call *back* to the original function's *concrete* body. So it feels weird to use `augmented` to declare a function that you explicitly *can't* call back to using `augmented` when augmenting that exact function:

```dart
foo() augmented;

augment foo() {
  augmented; // Error!
}
```

But... I do like that it is clearly related to augmentations as a feature. I don't have any better ideas. So let's go with `'augmented' ';'` as another kind of `functionBody`.

Thoughts?

**Comment by TekExplorer** [comment]
What about aborted bodies?
```dart
final int? x =;
Foo foo() =>;
```
It's clear, it's concise, it's intuitive, you don't need to learn the meaning of a word in a new context, nor add a new keyword, and it's also just existing syntax with a minor grammar tweak.

It's completely unambiguous too, it _has_ to be augmented. If it isn't erroring, then you know the augmentation _has_ to exist. If it's erroring, you _know_ it doesn't.

**Comment by tatumizer** [comment]
how about
```dart
proto final int? x;
proto Foo foo();
class Foo {
  proto factory Foo.fromJson(json);
}
```

**Comment by TekExplorer** [comment]
> how about
> ```dart
> proto final int? x;
> proto Foo foo();
> class Foo {
>   proto factory Foo.fromJson(json);
> }
> ```

Another new arbitrary keyword? That seems entirely unnecessary, and feels super cluttered - especially since that word becomes useless for users who just want to see the interface, and that word provides no value to them.

If we _must_ use a word, it should be on the end.

**Comment by munificent** [comment]
> What about aborted bodies?
> 
> final int? x =;
> Foo foo() =>;
> class Foo {
>   factory Foo.fromJson(json) =>;
> }
> It's clear, it's concise, it's intuitive, you don't need to learn the meaning of a word in a new context, nor add a new keyword, and it's also just existing syntax with a minor grammar tweak.

I agree, technically it works. But I have to admit that it just looks wrong to me to have an arrow that points to nothing.

And, if we ever do optional semicolons (which I still hold *some* hope for), this syntax will then become catastrophically problematic.

**Comment by TekExplorer** [comment]
> I don't agree with your characterization of the feature as "useless". C# uses an even bulkier term "partial" in a similar sense.
> Without it, the reader will be wondering what the definition means. Sometimes it's not even clear if this is the whole definition or just a part of it (which is the topic of the current thread).

I suppose. I'm just wary of even more keywords.

> Syntax highlighter can de-emphasize the word by painting it in some shy color (grey or something).

True.

And to be fair, we do have `late`

**Comment by TekExplorer** [comment]
> > What about aborted bodies?
> > 
> > final int? x =;
> > Foo foo() =>;
> > class Foo {
> >   factory Foo.fromJson(json) =>;
> > }
> > It's clear, it's concise, it's intuitive, you don't need to learn the meaning of a word in a new context, nor add a new keyword, and it's also just existing syntax with a minor grammar tweak.
> 
> I agree, technically it works. But I have to admit that it just looks wrong to me to have an arrow that points to nothing.

That's kind of the point. You start reading it and
- you see the interface
  - most people don't need more than that
- you see the arrow, so it's not abstract
  - useful for those extending it, as it means the implementation is actually "here", and not in a superclass.
- huh? Nothing? Well, it can't be valid, so the actual code must be somewhere else.
  - oh? Is that an augmentation my tooling is informing me of? Don't mind if I check it out.

> And, if we ever do optional semicolons (which I still hold *some* hope for), this syntax will then become catastrophically problematic.

I don't really see the point to optional semicolons tbh. Complicating parsing makes the whole language slower to iterate, and suddenly you now have a new tier of ambiguity issues. Idk.

And also not really, since if semicolons were optional, the semicolon would be _required_ for the syntax

**Comment by munificent** [comment]
I'm going to close this because the initial question is resolved now that macros are out of the picture. For the question about whether members can ambiguously be abstract or incomplete, let's discuss that on #4363.
